libs/isograph-react/src/loadable-hooks/useSkipLimitPagination.ts

```diff
@@ -152,47 +174,93 @@ export function useSkipLimitPagination<
         },
       ];
 
+      Promise.all([
```

Comment:
`getFetchMore` is called only once for every result, so it makes sense to "await"  the initial result here, then push it to `readOutDataAndRecords` array. This fixes problem introduced with changing `areObjectEqualWithDeepEquality`.

Then we can subscribe all the `readOutDataAndRecords`, we don't have to slice `loadedReferences`.

Does this makes sense?

Reply:
As much as possible, it's best to avoid calling `setState` any time except synchronously in response to a user event. i.e. we should synchronously call `setState` when `fetchMore` is called.

Calling `setState` in a callback makes it really easy for bugs (race conditions) to sneak in.

So, as much as possible, we should limit that to framework code: stuff like `PromiseWrapper`, Isograph store subscriptions, etc. should handle all of the async setting of state. And these should either be simple enough to not need testing, or (eventually) be well tested.

That will also make it easier to avoid race conditions when we e.g. add error handling, automatic refetching, etc.

What you have here (`cloned = [...original]; cloned[i] = new`) looks correct.

------

So, we should call `setState` in the same tick that `fetchMore` is called and "unwrap" during render.

Naively unwrapping all of the fragment references gives us a suspenseful pagination hook (see [this](https://github.com/isographlabs/isograph/commit/7fbb7f13c59ad7377581bd3c9088d08cbdf583c0#diff-94897d429e727e10ebdebf89fb27b8bbe95084c6bdd41a91c501c28e6a29a6d9R76-R98))

The extra work required to return a disciminated instead of suspending is annoying! I'm not wed to it. I think once we finish the `Readable` PR, we can dump all of that complexity to the user! The user will just call `readAllInArray` or `readAllButLastOneInArray` or whatever.

Reply:
How about this to fix the rerender https://github.com/isographlabs/isograph/compare/main...PatrykWalach:fix-skip-limit-does-not-rerender? This simplifies the hook quite a bit. Could be landed as separate PR.

Reply:
Using ".then" was a quick and dirty way to do it, the better way was `startTransition` so I've tried it here https://github.com/isographlabs/isograph/compare/main...PatrykWalach:cc70dc0a530b7fbefb9ed95250f58f325974c229, and it worked well, but the `pendingFragment` was missing, so I used `useDeferredValue` instead.